attestation-agent/attester/src/tdx/mod.rs

```diff
@@ -5,22 +5,61 @@
 
 use crate::utils::pad;
 
+use super::tsm_report::*;
 use super::Attester;
 use anyhow::*;
 use base64::Engine;
 use log::debug;
 use scroll::Pread;
 use serde::{Deserialize, Serialize};
 use std::path::Path;
-use tdx_attest_rs::{self, tdx_report_t};
 
 mod report;
 
+#[cfg(feature = "tdx-getquote-ioctl")]
+use tdx_attest_rs::{self, tdx_report_t};
+
 const TDX_REPORT_DATA_SIZE: usize = 64;
 const CCEL_PATH: &str = "/sys/firmware/acpi/tables/data/CCEL";
 
+fn tdx_tsm_report_provider_is_available() -> bool {
+    match TsmReportPath::new(TsmReportProvider::Tdx) {
+        Result::Err(_e) => false,
+        Result::Ok(_) => true,
+    }
+}
+
+fn tdx_getquote_ioctl_is_available() -> bool {
+    if cfg!(feature = "tdx-getquote-ioctl") {
+        Path::new("/dev/tdx-attest").exists() || Path::new("/dev/tdx-guest").exists()
+    } else {
+        false
+    }
+}
+
+#[allow(unused_variables)]
+fn get_quote_ioctl(report_data: &Vec<u8>) -> Result<Vec<u8>> {
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "tdx-getquote-ioctl")] {
+            let tdx_report_data = tdx_attest_rs::tdx_report_data_t {
+                // report_data.resize() ensures copying report_data to
+                // tdx_attest_rs::tdx_report_data_t cannot panic.
+                d: report_data.as_slice().try_into().unwrap(),
+            };
+
+            match tdx_attest_rs::tdx_att_get_quote(Some(&tdx_report_data), None, None, 0) {
+                (tdx_attest_rs::tdx_attest_error_t::TDX_ATTEST_SUCCESS, Some(q)) => Ok(q),
+                (error_code, _) => Err(anyhow!("TDX getquote ioctl also failed (error code: {:?}) after:", error_code)),
+            }
+        } else {
+            // return empty error
+            Err(anyhow!(""))
```

Comment:
Should this be something like "TDX Ioctl not supported?" Also, can you move this to the top of the function inside one small cfg_if?

Reply:
> Also, can you move this to the top of the function inside one small cfg_if?

I don't understand what to do here. you mean have two separate cfg_ifs?

Reply:
I was thinking you could do it with just one, but actually I guess you need to avoid using the ioctl deps when they are not enabled.